### PR TITLE
[#1235] ported arithmetics example to use java instead of xtend

### PR DESCRIPTION
**Pull request by cdietrich**
[#1235] ported arithmetics example to use java instead of xtend
Signed-off-by: Christian Dietrich <christian.dietrich@itemis.de>

### CODE REVIEW

File: org.eclipse.xtext.xtext.ui.examples/projects/arithmetics/org.eclipse.xtext.example.arithmetics.ui/src/org/eclipse/xtext/example/arithmetics/ui/autoedit/InterpreterAutoEdit.java [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+	@Override
+	public void customizeDocumentCommand(IDocument document, DocumentCommand command) {
+		for (String lineDelimiter : document.getLegalLineDelimiters()) {
+			if (Objects.equal(command.text, lineDelimiter)) {
+				try {
+					int line = document.getLineOfOffset(command.offset);
+					int lineStart = document.getLineOffset(line);
+					String text = document.get(lineStart, 3);
+					if (!Objects.equal(text, "def")) {
```
Comment (szarnekow) [inline org.eclipse.xtext.xtext.ui.examples/projects/arithmetics/org.eclipse.xtext.example.arithmetics.ui/src/org/eclipse/xtext/example/arithmetics/ui/autoedit/InterpreterAutoEdit.java]: `"def".equals(text)`

File: org.eclipse.xtext.xtext.ui.examples/projects/arithmetics/org.eclipse.xtext.example.arithmetics.ui/src/org/eclipse/xtext/example/arithmetics/ui/autoedit/InterpreterAutoEdit.java [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+						BigDecimal computedResult = computeResult(document, command);
+						if ((computedResult != null)) {
+							command.text = lineDelimiter + "// = " + computedResult + lineDelimiter;
+						}
+					}
+				} catch (Throwable t) {
+					if (t instanceof BadLocationException) {
+					} else {
+						throw Exceptions.sneakyThrow(t);
```
Comment (szarnekow) [inline org.eclipse.xtext.xtext.ui.examples/projects/arithmetics/org.eclipse.xtext.example.arithmetics.ui/src/org/eclipse/xtext/example/arithmetics/ui/autoedit/InterpreterAutoEdit.java]: I prefer this without sneaky throws

File: org.eclipse.xtext.xtext.ui.examples/projects/arithmetics/org.eclipse.xtext.example.arithmetics.ui/src/org/eclipse/xtext/example/arithmetics/ui/autoedit/InterpreterAutoEdit.java [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+		});
+	}
+
+	protected BigDecimal evaluate(Evaluation stmt) {
+		return new Calculator().evaluate(stmt.getExpression());
+	}
+
+	protected Evaluation findEvaluation(DocumentCommand command, XtextResource state) {
+		if (!state.getContents().isEmpty()) {
+			org.eclipse.xtext.example.arithmetics.arithmetics.Module m = IterableExtensions.<org.eclipse.xtext.example.arithmetics.arithmetics.Module>head(
```
Comment (szarnekow) [inline org.eclipse.xtext.xtext.ui.examples/projects/arithmetics/org.eclipse.xtext.example.arithmetics.ui/src/org/eclipse/xtext/example/arithmetics/ui/autoedit/InterpreterAutoEdit.java]: maybe use `state.getContents().get(0)` instead?

File: org.eclipse.xtext.xtext.ui.examples/projects/arithmetics/org.eclipse.xtext.example.arithmetics.ui/src/org/eclipse/xtext/example/arithmetics/ui/contentassist/ArithmeticsProposalProvider.java [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+	 * sets the selection on the arguments for proposed functions.
+	 */
+	@Override
+	public void completePrimaryExpression_Func(EObject model, Assignment assignment,
+			ContentAssistContext context, ICompletionProposalAcceptor acceptor) {
+		super.completePrimaryExpression_Func(model, assignment, context,
+				new ICompletionProposalAcceptor.Delegate(acceptor) {
+					@Override
+					public void accept(ICompletionProposal proposal) {
+						if ((proposal instanceof ConfigurableCompletionProposal)) {
```
Comment (szarnekow) [inline org.eclipse.xtext.xtext.ui.examples/projects/arithmetics/org.eclipse.xtext.example.arithmetics.ui/src/org/eclipse/xtext/example/arithmetics/ui/contentassist/ArithmeticsProposalProvider.java]: Unnecessary parenthesis

File: org.eclipse.xtext.xtext.ui.examples/projects/arithmetics/org.eclipse.xtext.example.arithmetics.ui/src/org/eclipse/xtext/example/arithmetics/ui/labeling/ArithmeticsLabelProvider.java [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+	public ArithmeticsLabelProvider(AdapterFactoryLabelProvider delegate) {
+		super(delegate);
+	}
+
+	public String text(org.eclipse.xtext.example.arithmetics.arithmetics.Module ele) {
+		return ele.getName();
+	}
+
+	public String text(Definition ele) {
+		return ele.getName() + IterableExtensions.join(ele.getArgs(), "(", ",", ")", (p) -> {
```
Comment (szarnekow) [inline org.eclipse.xtext.xtext.ui.examples/projects/arithmetics/org.eclipse.xtext.example.arithmetics.ui/src/org/eclipse/xtext/example/arithmetics/ui/labeling/ArithmeticsLabelProvider.java]: Lambda can be simplified

File: org.eclipse.xtext.xtext.ui.examples/projects/arithmetics/org.eclipse.xtext.example.arithmetics.ui/src/org/eclipse/xtext/example/arithmetics/ui/quickfix/ArithmeticsQuickfixProvider.java [hunk 6]
```diff
@@ -17,17 +17,19 @@
...
-      it.getXtextDocument().replace((issue.getOffset()).intValue(), (issue.getLength()).intValue(), string);
-    };
-    acceptor.accept(issue, ("Replace with " + string), (("Replace expression with \'" + string) + "\'"), "upcase.png", _function);
-  }
+	@Fix(ArithmeticsValidator.NORMALIZABLE)
+	public void normalize(Issue issue, IssueResolutionAcceptor acceptor) {
+		String string = issue.getData()[0];
+		acceptor.accept(issue, "Replace with " + string, "Replace expression with \'" + string + "\'", "upcase.png",
+				(IModification) (IModificationContext it) -> {
+					it.getXtextDocument().replace((issue.getOffset()).intValue(), (issue.getLength()).intValue(),
```
Comment (szarnekow) [inline org.eclipse.xtext.xtext.ui.examples/projects/arithmetics/org.eclipse.xtext.example.arithmetics.ui/src/org/eclipse/xtext/example/arithmetics/ui/quickfix/ArithmeticsQuickfixProvider.java]: Unnecessary explicit unboxing

File: org.eclipse.xtext.xtext.ui.examples/projects/arithmetics/org.eclipse.xtext.example.arithmetics/src-gen/org/eclipse/xtext/example/arithmetics/arithmetics/Import.java [hunk 7]
```diff
@@ -38,12 +38,12 @@
    * </p>
    * <!-- end-user-doc -->
    * @return the value of the '<em>Module</em>' reference.
-   * @see #setModule(org.eclipse.xtext.example.arithmetics.arithmetics.Module)
+   * @see #setModule(Module)
```
Comment (szarnekow) [inline org.eclipse.xtext.xtext.ui.examples/projects/arithmetics/org.eclipse.xtext.example.arithmetics/src-gen/org/eclipse/xtext/example/arithmetics/arithmetics/Import.java]: Local package has higher precedence than java.lang?

File: org.eclipse.xtext.xtext.ui.examples/projects/arithmetics/org.eclipse.xtext.example.arithmetics/src-gen/org/eclipse/xtext/example/arithmetics/arithmetics/impl/AbstractDefinitionImpl.java [hunk 8]
```diff
@@ -174,7 +174,7 @@ public String toString()
   {
     if (eIsProxy()) return super.toString();
 
-    StringBuilder result = new StringBuilder(super.toString());
+    StringBuffer result = new StringBuffer(super.toString());
```
Comment (szarnekow) [inline org.eclipse.xtext.xtext.ui.examples/projects/arithmetics/org.eclipse.xtext.example.arithmetics/src-gen/org/eclipse/xtext/example/arithmetics/arithmetics/impl/AbstractDefinitionImpl.java]: This looks like an old emf version to me.

File: org.eclipse.xtext.xtext.ui.examples/projects/arithmetics/org.eclipse.xtext.example.arithmetics/src/org/eclipse/xtext/example/arithmetics/generator/ArithmeticsGenerator.java [hunk 9]
```diff
@@ -17,9 +17,18 @@
...
  */
-@SuppressWarnings("all")
 public class ArithmeticsGenerator extends AbstractGenerator {
-  @Override
-  public void doGenerate(final Resource resource, final IFileSystemAccess2 fsa, final IGeneratorContext context) {
-  }
+
+	@Override
+	public void doGenerate(Resource resource, IFileSystemAccess2 fsa, IGeneratorContext context) {
+//		Iterator<Greeting> filtered = Iterators.filter(resource.getAllContents(), Greeting.class);
```
Comment (szarnekow) [inline org.eclipse.xtext.xtext.ui.examples/projects/arithmetics/org.eclipse.xtext.example.arithmetics/src/org/eclipse/xtext/example/arithmetics/generator/ArithmeticsGenerator.java]: Should be removed.

File: org.eclipse.xtext.xtext.ui.examples/projects/arithmetics/org.eclipse.xtext.example.arithmetics/src/org/eclipse/xtext/example/arithmetics/validation/ArithmeticsValidator.java [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+
+	public static final String NORMALIZABLE = "normalizable-expression";
+
+	@Inject
+	private Calculator calculator;
+
+	@Check
+	public void checkDivByZero(Div div) {
+		double value = calculator.evaluate(div.getRight()).doubleValue();
+		if (value == 0.0) {
```
Comment (szarnekow) [inline org.eclipse.xtext.xtext.ui.examples/projects/arithmetics/org.eclipse.xtext.example.arithmetics/src/org/eclipse/xtext/example/arithmetics/validation/ArithmeticsValidator.java]: Double.compare(value, 0.0) == 0

File: org.eclipse.xtext.xtext.ui.examples/projects/arithmetics/org.eclipse.xtext.example.arithmetics/src/org/eclipse/xtext/example/arithmetics/validation/ArithmeticsValidator.java [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+			error("Division by zero detected.", ArithmeticsPackage.Literals.DIV__RIGHT);
+		}
+	}
+
+	@Check
+	public void checkNormalizable(Expression expr) {
+		if (expr instanceof NumberLiteral || expr instanceof FunctionCall) {
+			return;
+		}
+		Evaluation eval = EcoreUtil2.<Evaluation>getContainerOfType(expr, Evaluation.class);
```
Comment (szarnekow) [inline org.eclipse.xtext.xtext.ui.examples/projects/arithmetics/org.eclipse.xtext.example.arithmetics/src/org/eclipse/xtext/example/arithmetics/validation/ArithmeticsValidator.java]: Unnecessary type arg

File: org.eclipse.xtext.xtext.ui.examples/projects/arithmetics/org.eclipse.xtext.example.arithmetics/src-gen/org/eclipse/xtext/example/arithmetics/arithmetics/Import.java (hunk 7 above)
Comment (cdietrich) [reply org.eclipse.xtext.xtext.ui.examples/projects/arithmetics/org.eclipse.xtext.example.arithmetics/src-gen/org/eclipse/xtext/example/arithmetics/arithmetics/Import.java]: no idea what java does here will fix back

Comment (cdietrich) [reply org.eclipse.xtext.xtext.ui.examples/projects/arithmetics/org.eclipse.xtext.example.arithmetics/src-gen/org/eclipse/xtext/example/arithmetics/arithmetics/Import.java]: maybe this is cause of inconsistent emf versions.
time to do that damn emf update

File: org.eclipse.xtext.xtext.ui.examples/projects/arithmetics/org.eclipse.xtext.example.arithmetics/src-gen/org/eclipse/xtext/example/arithmetics/arithmetics/impl/AbstractDefinitionImpl.java (hunk 8 above)
Comment (cdietrich) [reply org.eclipse.xtext.xtext.ui.examples/projects/arithmetics/org.eclipse.xtext.example.arithmetics/src-gen/org/eclipse/xtext/example/arithmetics/arithmetics/impl/AbstractDefinitionImpl.java]: i dont know how this was generated before

**Comment by cdietrich** [comment]
i dont know how that other emf version ever landed there. 
we still suffer from never working on the emf update

**Comment by cdietrich** [comment]
have no idea how emf 2.14 ever landed here.

File: org.eclipse.xtext.xtext.ui.examples/projects/arithmetics/org.eclipse.xtext.example.arithmetics.ui/src/org/eclipse/xtext/example/arithmetics/ui/outline/ArithmeticsOutlineTreeProvider.java [hunk 12]
```diff
@@ -1,33 +1,32 @@
...
-    };
-    Iterables.<Definition>filter(module.eContents(), Definition.class).forEach(_function);
-  }
-  
-  public boolean _isLeaf(final Definition definition) {
-    return true;
-  }
+
+	public void _createChildren(IOutlineNode parentNode,
+			org.eclipse.xtext.example.arithmetics.arithmetics.Module module) {
```
Comment (tivervac) [inline org.eclipse.xtext.xtext.ui.examples/projects/arithmetics/org.eclipse.xtext.example.arithmetics.ui/src/org/eclipse/xtext/example/arithmetics/ui/outline/ArithmeticsOutlineTreeProvider.java]: import the FQN

File: org.eclipse.xtext.xtext.ui.examples/projects/arithmetics/org.eclipse.xtext.example.arithmetics.ui/src/org/eclipse/xtext/example/arithmetics/ui/autoedit/InterpreterAutoEdit.java [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+		});
+	}
+
+	protected BigDecimal evaluate(Evaluation stmt) {
+		return new Calculator().evaluate(stmt.getExpression());
+	}
+
+	protected Evaluation findEvaluation(DocumentCommand command, XtextResource state) {
+		if (!state.getContents().isEmpty()) {
+			org.eclipse.xtext.example.arithmetics.arithmetics.Module m =
```
[the remark this answers is not in the document]
Comment (cdietrich) [reply org.eclipse.xtext.xtext.ui.examples/projects/arithmetics/org.eclipse.xtext.example.arithmetics.ui/src/org/eclipse/xtext/example/arithmetics/ui/autoedit/InterpreterAutoEdit.java]: might be a java 11 problem

File: org.eclipse.xtext.xtext.ui.examples/projects/arithmetics/org.eclipse.xtext.example.arithmetics.ui/src/org/eclipse/xtext/example/arithmetics/ui/contentassist/ArithmeticsProposalProvider.java [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+							}
+						}
+						super.accept(proposal);
+					}
+				});
+	}
+
+	@Override
+	protected String getDisplayString(EObject element, String qualifiedNameAsString,
+			String shortName) {
```
[the remark this answers is not in the document]
Comment (cdietrich) [reply org.eclipse.xtext.xtext.ui.examples/projects/arithmetics/org.eclipse.xtext.example.arithmetics.ui/src/org/eclipse/xtext/example/arithmetics/ui/contentassist/ArithmeticsProposalProvider.java]: no idea. am using formatter

**Pull request merged**